xdsl/transforms/csl_stencil_bufferize.py

```diff
@@ -371,13 +379,77 @@ def match_and_rewrite(self, op: arith.Constant, rewriter: PatternRewriter, /):
         )
 
 
+@dataclass(frozen=True)
+class ReselectLinalgOutsFromInputs(RewritePattern):
+    """
+    Reselects the DPS `outs` of a linalg op if it is one of its inputs.
+      * select `writable` tensor input with no later use, or else
+      * select linalg op input with no later use
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(
+        self, op: linalg.NamedOpBase | linalg.Generic, rewriter: PatternRewriter, /
+    ):
+        # only apply rewrite when re-selecting `outs` from `ins`
+        if (
+            op.outputs[0] not in op.inputs
+            or self.is_writable(op.outputs[0])
+            and len(op.outputs) == 1
+        ):
+            return
+
+        # the new `outs` to re-select
+        out: SSAValue | None = None
+
+        for arg in op.inputs:
+            # reselect outs that has no later use to avoid read-after-write conflicts
+            if len(arg.uses) == 1:
```

Comment:
Does this check for later uses or all uses including ones before this one?

E.g.

```mlir
%0 = ...
%1 = "test.op"() : () -> memref<10xf32>
%2 = ...
"test.op"(%1) : (memref<10xf32>) -> ()
linalg.add ins(%1, %2 : memref<10xf32>, memref<10xf32>) outs(%0 : memref<10xf32>)

// %1 not used after here, but len(uses) > 1?
```

Reply:
It checks for no other uses at all, as the check for no later uses isn't quite straightforward to implement. Both should logically work, though in theory there might be some additional edge cases that might be caught checking for later uses.